File: cli/src/preflights/preflight-add.ts
import chalk from "chalk";
import fs from "fs-extra";
import { join } from "path";
import { Component, getComponent } from "../utils/registry.js";
import { isAlreadyInitialized, validateFlutterProject } from "../utils/validator.js";

const { existsSync } = fs;

export interface PreflightResult {
  success: boolean;
  error?: string;
  details?: string[];
  component?: Component;
}

/**
 * Pre-flight checks for the add command
 * @param cwd - Current working directory
 * @param componentName - Name of the component to add
 * @returns PreflightResult with validation status and messages
 */
export function preflightAdd(cwd: string, componentName: string): PreflightResult {
  // Step 1: Validate Flutter project
  const validation = validateFlutterProject(cwd);

  if (!validation.isValid) {
    return {
      success: false,
      error: chalk.red("✗ Flutter project validation failed"),
      details: [chalk.gray(`  • ${validation.error}`)],
    };
  }

  // Step 2: Check if fluttercn is initialized
  if (!isAlreadyInitialized(cwd)) {
    return {
      success: false,
      error: chalk.red("✗ fluttercn is not initialized"),
      details: [
        chalk.gray("  Please run ") +
          chalk.cyan("fluttercn init") +
          chalk.gray(" first to set up fluttercn."),
      ],
    };
  }

  // Step 3: Check if component exists
  const component = getComponent(componentName);

  if (!component) {
    return {
      success: false,
      error: chalk.red(`✗ Component '${chalk.yellow(componentName)}' not found`),
      details: [
        chalk.gray("  Run ") +
          chalk.cyan("fluttercn list") +
          chalk.gray(" to see available components."),
      ],
    };
  }

  // Step 4: Check dependencies
  if (component.dependencies && component.dependencies.length > 0) {
    const missingDeps: string[] = [];

    for (const dep of component.dependencies) {
      const depPath = join(cwd, "lib", dep.file);
      if (!existsSync(depPath)) {
        missingDeps.push(dep.message || dep.file);
      }
    }

    if (missingDeps.length > 0) {
      return {
        success: false,
        error: chalk.red("✗ Missing required dependencies"),
        details: missingDeps.map((dep) => chalk.gray(`  • ${dep}`)),
      };
    }
  }

  return {
    success: true,
    component,
  };
}

File: dart-cli/lib/preflights/preflight_add.dart
import 'dart:io';

import 'package:fluttercn/utils/registry.dart';
import 'package:fluttercn/utils/validator.dart';
import 'package:path/path.dart' as path;

/// Preflight result with success status and optional component data
class PreflightResult {
  final bool success;
  final String? error;
  final List<String>? details;
  final Component? component;

  PreflightResult({
    required this.success,
    this.error,
    this.details,
    this.component,
  });
}

PreflightResult preflightAdd(String cwd, String componentName) {
  final validation = validateFlutterProject(cwd);

  if (!validation.isValid) {
    return PreflightResult(
      success: false,
      error: '✗ Flutter project validation failed',
      details: ['  • ${validation.error}'],
    );
  }

  if (!isAlreadyInitialized(cwd)) {
    return PreflightResult(
      success: false,
      error: '✗ fluttercn is not initialized',
      details: [
        '  Please run fluttercn init first to set up fluttercn.',
      ],
    );
  }

  final component = getComponent(componentName);

  if (component == null) {
    return PreflightResult(
      success: false,
      error: "✗ Component '$componentName' not found",
      details: [
        '  Run fluttercn list to see available components.',
      ],
    );
  }

  if (component.dependencies != null && component.dependencies!.isNotEmpty) {
    final missingDeps = <String>[];

    for (final dep in component.dependencies!) {
      final depPath = path.join(cwd, 'lib', dep.file);
      if (!File(depPath).existsSync()) {
        missingDeps.add(dep.message ?? dep.file);
      }
    }

    if (missingDeps.isNotEmpty) {
      return PreflightResult(
        success: false,
        error: '✗ Missing required dependencies',
        details: missingDeps.map((dep) => '  • $dep').toList(),
      );
    }
  }

  return PreflightResult(
    success: true,
    component: component,
  );
}
